util/containerutil/stub.go

```diff
@@ -3,16 +3,30 @@ package containerutil
 import (
 	"context"
 	"io"
+
+	"github.com/pkg/errors"
 )
 
 // This is a stub for use in internal testing when its too much effort to provide a legitimate backend.
 // Should never be used IRL.
-type stubFrontend struct{}
+type stubFrontend struct {
+	*shellFrontend
+}
 
 // NewStubFrontend creates a stubbed frontend. Useful in cases where a frontend could not be detected, but we still need a frontend.
 // Examples include earthly/earthly, or integration tests. It is currently only used as a fallback when docker or other frontends are missing.
-func NewStubFrontend(ctx context.Context) (ContainerFrontend, error) {
-	return &stubFrontend{}, nil
+func NewStubFrontend(ctx context.Context, cfg *FrontendConfig) (ContainerFrontend, error) {
+	fe := &stubFrontend{
+		shellFrontend: &shellFrontend{},
+	}
+
+	var err error
+	fe.urls, err = fe.setupAndValidateAddresses(FrontendStub, cfg)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to calculate buildkit URLs")
+	}
```

Comment:
This was the source of the bug. We used to calculate URLs after choosing a frontend; but by choosing the stub, we would end up skipping the calculations instead. The solution was to also let the stub calculate URLs; so the remote buildkit would end up being chosen.

Reply:
Ohhhh buddy did this take a while to figure out.